File: src/main/golang/leetcode/editor/cn/[28]Find the Index of the First Occurrence in a String.go
package lc

//leetcode submit region begin(Prohibit modification and deletion)
func strStr(haystack string, needle string) int {
	//n := len(needle)
	//var window []rune
	//for i, char := range haystack {
	//	window = append(window, char)
	//	if i >= n-1 {
	//		str := string(window)
	//		if str == needle {
	//			return i - n + 1
	//		}
	//		window = window[1:]
	//	}
	//}
	//return -1

	build := func(s string) [][256]int {
		n := len(s)
		dp := make([][256]int, n)
		dp[0][s[0]] = 1
		var x int
		for i := 1; i < n; i++ {
			for k := 0; k < 256; k++ {
				dp[i][k] = dp[x][k]
			}
			dp[i][s[i]] = i + 1
			x = dp[x][s[i]]
		}
		return dp
	}
	dp := build(needle)
	n := len(needle)
	var j int
	for i := 0; i < len(haystack); i++ {
		c := haystack[i]
		j = dp[j][c]
		if j == n {
			return i - n + 1
		}
	}
	return -1
}

//leetcode submit region end(Prohibit modification and deletion)

File: src/main/golang/leetcode/editor/cn/[28]Find the Index of the First Occurrence in a String.py
# leetcode submit region begin(Prohibit modification and deletion)
class Solution:
    # def strStr(self, haystack: str, needle: str) -> int:
    #     try:
    #         return haystack.index(needle)
    #     except:
    #         return -1
    def strStr(self, haystack: str, needle: str) -> int:
        n, m = len(haystack), len(needle)
        for i in range(n - m + 1):
            if haystack[i] == needle[0]:
                j = 0
                while j < m:
                    if needle[j] != haystack[i + j]:
                        break
                    j += 1
                if j == m: return i
        return -1

# leetcode submit region end(Prohibit modification and deletion)
